openjdk/mmtkBarrierSetC2.cpp

```diff
@@ -593,3 +593,43 @@ void MMTkBarrierSetC2::expand_allocate(PhaseMacroExpand* x,
   x->transform_later(result_phi_i_o);
   // This completes all paths into the result merge point
 }
+
+bool MMTkBarrierSetC2::can_remove_barrier(GraphKit* kit, PhaseTransform* phase, Node* src, Node* slot, Node* val, bool skip_const_null) const {
+  // Skip barrier if the new target is a null pointer.
+  if (skip_const_null && val != NULL && val->is_Con() && val->bottom_type() == TypePtr::NULL_PTR) {
+    return true;
+  }
+  // Barrier elision based on allocation node does not working well with slowpath-only allocation.
+  if (!MMTK_ENABLE_ALLOCATION_FASTPATH) return false;
+  // No barrier required for newly allocated objects.
+  if (src == kit->just_allocated_object(kit->control())) return true;
```

Comment:
This is generic to all barriers with young optimization. LXR uses this as well. If a barrier wants to track young objects, it can override this method and switch to their own barrier elision policy.